File: MiniMax/MiniMax_TicTacToe_Lib/TicTacToeLib.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace MiniMax_TicTacToe_Lib
{
    public enum PieceType : sbyte { X, O, Empty }

    public class TicTacToeLib
    {
        private readonly PieceType Player;

        #region Constructor
        public TicTacToeLib(PieceType player)
        {
            Player = player;
        }
        #endregion

        public void HumanvsHuman(PieceType startPiece)
        {
            var currentPiece = startPiece;
            var board = CreateEmptyBoard();
            PrintBoard(board);
            var column = 0;
            var row = 0;

            while (column != -1)
            {
                Console.WriteLine("{0}'s move", currentPiece);

                Console.WriteLine("Enter the row you want to drop a piece");
                row = int.Parse(Console.ReadLine());
                Console.WriteLine("Enter the column you want to drop a piece");
                column = int.Parse(Console.ReadLine());

                board[row][column] = currentPiece;

                if (column != -1)
                {
                    PrintBoard(board);

                    if (CheckWin(board))
                    {
                        Console.WriteLine("You Won, {0}!", currentPiece);
                        return;
                    }
                    currentPiece = (currentPiece == PieceType.O) ? PieceType.X : PieceType.O;
                }
            }
        }

        public void HumanVsComputer(PieceType player, int depth)
        {
            var board = CreateEmptyBoard();
            PrintBoard(board);
            var column = 0;
            var row = 0;
            while (column != -1)
            {
                Console.WriteLine("Enter the row you want to drop a piece");
                row = int.Parse(Console.ReadLine());
                Console.WriteLine("Enter the column you want to drop a piece");
                column = int.Parse(Console.ReadLine());

                board[row][column] = player;

                PrintBoard(board);

                if (CheckWin(board))
                {
                    Console.WriteLine("You Won!");
                    return;
                }

                //make pc move
                board = AI(board, player, depth);
                Console.WriteLine("AI Move:");
                PrintBoard(board);

                if (CheckWin(board))
                {
                    Console.WriteLine("The Pc Won!");
                    return;
                }
            }
        }

        public bool CheckWin(PieceType[][] board)
        {
            var top = board[0][0] == board[0][1] && board[0][1] == board[0][2] && board[0][2] != PieceType.Empty;
            var middle = board[1][0] == board[1][1] && board[1][1] == board[1][2] && board[1][2] != PieceType.Empty;
            var bottom = board[2][0] == board[2][1] && board[2][1] == board[2][2] && board[2][2] != PieceType.Empty;

            var left = board[0][0] == board[1][0] && board[1][0] == board[2][0] && board[2][0] != PieceType.Empty;
            var center = board[0][1] == board[1][1] && board[1][1] == board[2][1] && board[2][1] != PieceType.Empty;
            var right = board[0][2] == board[1][2] && board[1][2] == board[2][2] && board[2][2] != PieceType.Empty;

            var TLBRDiag = board[0][0] == board[1][1] && board[1][1] == board[2][2] && board[2][2] != PieceType.Empty;
            var BLTRDiag = board[2][0] == board[1][1] && board[1][1] == board[0][2] && board[0][2] != PieceType.Empty;

            return top || middle || bottom || left || center || right || TLBRDiag || BLTRDiag;
        }

        #region Board Operations
        public void PrintBoard(PieceType[][] board)
        {
            Console.Clear();
            //created to maintain default system console color
            Console.ForegroundColor = ConsoleColor.Black;
            Console.BackgroundColor = ConsoleColor.White;
            for (int i = 0; i < board.Length; i++)
            {
                for (int j = 0; j < board[i].Length; j++)
                {
                    var output = ' ';
                    if (board[i][j] == PieceType.X)
                    {
                        output = 'X';
                        Console.ForegroundColor = ConsoleColor.Red;
                    }
                    else if (board[i][j] == PieceType.O)
                    {
                        output = 'O';
                        Console.ForegroundColor = ConsoleColor.Blue;
                    }
                    Console.Write(" " + output);
                    Console.ForegroundColor = ConsoleColor.Black;
                    if (j != board[i].Length-1)
                    {
                        Console.Write("|");
                    }
                }
                Console.WriteLine();
                if (i != board.Length - 1)
                {
                    Console.WriteLine("--------");
                }
            }
            Console.WriteLine();
        }

        public PieceType[][] CopyBoard(PieceType[][] board)
        {
            var newBoard = new PieceType[board.Length][];
            for (int i = 0; i < board.Length; i++)
            {
                newBoard[i] = new PieceType[board[i].Length];
                for (int j = 0; j < board[i].Length; j++)
                {
                    newBoard[i][j] = board[i][j];
                }
            }
            return newBoard;
        }

        public PieceType[][] CreateEmptyBoard()
        {
            var boardHeight = 3;

            var tempBoard = new PieceType[boardHeight][];

            for (int i = 0; i < boardHeight; i++)
            {
                tempBoard[i] = new PieceType[boardHeight];

                for (int j = 0; j < tempBoard[i].Length; j++)
                {
                    tempBoard[i][j] = PieceType.Empty;
                }
            }
            return tempBoard;
        }

        public Tuple<int, int>[] AvailableMoves(PieceType[][] board)
        {
            var available = new List<Tuple<int, int>>();

            for (int i = 0; i < board.Length; i++)
            {
                for (int j = 0; j < board[i].Length; j++)
                {
                    if (board[i][j] == PieceType.Empty)
                    {
                        available.Add(new Tuple<int, int>(i, j));
                    }
                }
            }
            return available.ToArray();
        }

        #endregion

        public int GradeState(Node state, PieceType piece, int depth)
        {
            //if a win
            if (CheckWin(state.Board))
            {

                //if the player is the winner the ai grades the state as a loss
                if (Player == piece)
                {
                    return -10; //depth;
                }
                //if the ai is the winner the ai grades the state as a win
                return 10; //-depth;
            }

            //TODO TEMPORARY TODO
            //draw
            return 0;
        }

        #region AI
        public Node GenerateMoveTree(PieceType[][] board, PieceType piece, int depth)
        {
            var outNode = new Node(board, piece);

            //checking for win, and stopping here if so.
            if (depth < 0 || CheckWin(board))
            {
                return outNode;
            }

            var availableMoves = AvailableMoves(board);
            foreach (var indices in availableMoves)
            {
                var newBoard = CopyBoard(board);
                PieceType oppositePiece = (piece == PieceType.O) ? PieceType.X : PieceType.O;
                newBoard[indices.Item1][indices.Item2] = oppositePiece;

                var node = GenerateMoveTree(newBoard, oppositePiece, depth - 1);
                outNode.Children.Add(node);
            }

            return outNode;
        }

        public int AlphaBetaMinMax(Node node, bool maxamizingPlayer, int depthScore, int alpha = int.MinValue, int beta = int.MaxValue)
        {
            if (node.Children.Count == 0)
            {
                return GradeState(node, node.UpdatedPiece, depthScore);
            }
            if (maxamizingPlayer)
            {
                var bestValue = int.MinValue;
                foreach (var child in node.Children)
                {
                    var grade = AlphaBetaMinMax(child, false, depthScore - 1, alpha, beta);
                    bestValue = Math.Max(bestValue, grade);
                    alpha = Math.Max(alpha, bestValue);
                    if (beta <= alpha)
                    {
                        break;
                    }
                }
                return bestValue;
            }
            else
            {
                var bestValue = int.MaxValue;
                foreach (var child in node.Children)
                {
                    var grade = AlphaBetaMinMax(child, true, depthScore - 1, alpha, beta);
                    bestValue = Math.Min(bestValue, grade);
                    beta = Math.Min(beta, bestValue);
                    if (beta <= alpha)
                    {
                        break;
                    }
                }
                return bestValue;
            }

        }

        public PieceType[][] AI(PieceType[][] board, PieceType piece, int depth)
        {
            var tree = GenerateMoveTree(board, piece, depth);
            foreach (var child in tree.Children)
            {
                child.Value = AlphaBetaMinMax(child, false, depth);
            }

            var maxScore = int.MinValue;
            var index = -1;
            for (int i = 0; i < tree.Children.Count; i++)
            {
                if (tree.Children[i].Value > maxScore)
                {
                    index = i;
                    maxScore = tree.Children[i].Value;
                }
            }
            return tree.Children[index].Board;
        }
        #endregion
    }
}

File: Neural/TicTacToeLib.py
from Constants import PieceType, bcolors
from Node import Node
import os


class TicTacToeLib:
    player = PieceType.Empty

    def HumanVsHuman(self, player_piece):
        self.player = player_piece
        current_piece = self.player
        board = self.create_empty_board()
        self.print_board(board)

        col = 0
        row = 0

        while col != -1:
            print("{0}'s move".format(current_piece))

            row = int(input('Enter the row you want to place a piece'))
            col = int(input('Enter the column you want to place a piece (-1 to quit)'))

            if col != -1:
                board[row][col] = current_piece

                self.print_board(board)

                if self.check_win(board):
                    print('You won {0}!'.format(current_piece))
                    return
                current_piece = self.invert_piece(current_piece)

    def HumanVsComputer(self, player_piece, depth):
        self.player = player_piece

        board = self.create_empty_board()
        self.print_board(board)
        col = 0
        row = 0

        while col != -1:
            row = int(input('Enter the row you want to place a piece'))
            col = int(input('Enter the col you want to place a piece (-1 to exit)'))

            board[row][col] = self.player

            self.print_board(board)

            if self.check_win(board):
                print('You won!')
                return

            if not self.board_has_moves(board):
                print('Board Full, restarting..')
                board = self.create_empty_board()
                continue

            board = self.Ai(board, depth)
            print('Ai Move:')
            self.print_board(board)

            if self.check_win(board):
                print('The PC won!')
                return

            if not self.board_has_moves(board):
                print('Board Full, restarting..')
                board = self.create_empty_board()
                continue

    def invert_piece(self, piece):
        return PieceType.X if (piece == PieceType.O) else PieceType.O

    def check_win(self, board):
        top = (board[0][0] == board[0][1]) and (board[0][1] == board[0][2]) and board[0][2] != PieceType.Empty
        middle = board[1][0] == board[1][1] and board[1][1] == board[1][2] and board[1][2] != PieceType.Empty
        bottom = board[2][0] == board[2][1] and board[2][1] == board[2][2] and board[2][2] != PieceType.Empty

        left = board[0][0] == board[1][0] and board[1][0] == board[2][0] and board[2][0] != PieceType.Empty
        center = board[0][1] == board[1][1] and board[1][1] == board[2][1] and board[2][1] != PieceType.Empty
        right = board[0][2] == board[1][2] and board[1][2] == board[2][2] and board[2][2] != PieceType.Empty

        TLBRDiag = board[0][0] == board[1][1] and board[1][1] == board[2][2] and board[0][2] != PieceType.Empty
        BLTRDiag = board[2][0] == board[1][1] and board[1][1] == board[0][2] and board[0][2] != PieceType.Empty

        return top or middle or bottom or left or center or right or TLBRDiag or BLTRDiag

    # board operations
    def print_board(self, board):
        for r in range(len(board)):
            line_string = ''
            for c in range(len(board[r])):
                output = ' '
                if board[r][c] == PieceType.X:
                    output = bcolors.WARNING + 'X' + bcolors.ENDC
                elif board[r][c] == PieceType.O:
                    output = bcolors.OKBLUE + 'O' + bcolors.ENDC
                line_string += ' ' + output
                if c != len(board[r])-1:
                    line_string += '|'
            print(line_string)
            if r != len(board) - 1:
                print('-'*7)

    def copy_board(self, board):
        out_board = []
        for row in range(len(board)):
            out_board.append([])
            for pos in range(len(board[row])):
                out_board[row].append(board[row][pos])
        return out_board

    def create_empty_board(self):
        return [[PieceType.Empty for x in range(3)] for x in range(3)]

    def get_available_moves(self, board):
        out_board = []
        for row in range(len(board)):
            for pos in range(len(board[row])):
                if board[row][pos] == PieceType.Empty:
                    out_board.append((row, pos))
        return out_board

    def board_has_moves(self, board):
        for row in board:
            for piece in row:
                if piece == PieceType.Empty:
                    return True
        return False

    def grade_state(self, node, piece, depth):
        # if a win
        if self.check_win(node.board):
            # if the player won, the ai grades the state as a loss
            if self.player == piece:
                return -10
            # if the ai is the winner, grade the state as a win
            return 10
        # its a draw
        return 0

    def generate_move_tree(self, board, piece, depth):

        outNode = Node(board, piece)

        if depth<0 or self.check_win(board):
            return outNode

        available_moves = self.get_available_moves(board)

        for r,c in available_moves:
            new_board = self.copy_board(board)
            op_piece = self.invert_piece(piece)
            new_board[r][c] = op_piece

            node = self.generate_move_tree(new_board, op_piece, depth-1)
            outNode.children.append(node)
        return outNode

    def alpha_beta_minmax(self, node, maximizing_player, depthScore, alpha = -10000, beta = 10000):
        if len(node.children) == 0:
            return self.grade_state(node, node.updated_piece, depthScore)
        if maximizing_player:
            best_value = -10000
            for child in node.children:
                grade = self.alpha_beta_minmax(child, False, depthScore - 1, alpha, beta)
                best_value = max(best_value, grade)
                alpha = max(alpha, best_value)
                if beta <= alpha:
                    break
            return best_value
        else:
            best_value = 10000
            for child in node.children:
                grade = self.alpha_beta_minmax(child, True, depthScore - 1, alpha, beta)
                best_value = min(best_value, grade)
                beta = min(beta, best_value)
                if beta <= alpha:
                    break
            return best_value

    def Ai(self, board, depth):
        tree = self.generate_move_tree(board, self.player, depth)
        for child in tree.children:
            child.value = self.alpha_beta_minmax(child, False, depth)

        max_score = -10000
        index = -1
        for i in range(len(tree.children)):
            if tree.children[i].value > max_score:
                index = i
                max_score = tree.children[i].value
        return tree.children[index].board
